containers/src/Data/Tree.hs

```diff
@@ -415,3 +456,23 @@ unfoldForestQ f aQ = case viewl aQ of
     splitOnto as (_:bs) q = case viewr q of
         q' :> a -> splitOnto (a:as) bs q'
         EmptyR -> error "unfoldForestQ"
+
+--------------------------------------------------------------------------------
+
+-- Note [Implemented Foldable Tree functions]
+-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+--
+-- Implemented:
+--
+-- foldMap, foldr, foldl': Basic functions.
+-- fold, elem: Implemented same as the default definition, but INLINABLE to
+-- allow specialization.
+-- foldr1, foldl1, null, maximum, minimum: Implemented more efficiently than
+-- defaults since trees are non-empty.
+-- sum, product: Implemented as strict left folds. Defaults use the lazy foldMap
+-- before base 4.15.1.
+--
+-- Not implemented:
+--
+-- foldMap', toList, length: Defaults perform well.
```

Comment:
I don't understand what makes `fold` and `elem` different from `foldMap'` in this regard. Are there issues specializing the first two but not the last? If so, do you have any guess why?

Reply:
The default definition `foldMap'` is perhaps small enough for GHC to inline without any help. But I don't know for sure.

Reply:
Bleh. GHC inlining pragmas are such a dark art. I hate it.